File: 09.KMPAlgo.java
/* ----------------------------------------------------------------------------  */
/*   ( The Authentic JS/JAVA CodeBuff )
 ___ _                      _              _ 
 | _ ) |_  __ _ _ _ __ _ __| |_ __ ____ _ (_)
 | _ \ ' \/ _` | '_/ _` / _` \ V  V / _` || |
 |___/_||_\__,_|_| \__,_\__,_|\_/\_/\__,_|/ |
                                        |__/ 
 */
/* --------------------------------------------------------------------------   */
/*    Youtube: https://youtube.com/@code-with-Bharadwaj                        */
/*    Github : https://github.com/Manu577228                                  */
/*    Portfolio : https://manu-bharadwaj-portfolio.vercel.app/portfolio        */
/* -----------------------------------------------------------------------  */

import java.util.*;

public class KMPAlgorithm {

    // Function to build LPS (Longest Prefix Suffix) array
    static int[] buildLPS(String p) {
        int m = p.length();        // length of pattern
        int[] lps = new int[m];    // lps array initialized to 0
        int len = 0;               // current length of longest prefix suffix
        int i = 1;                 // start from index 1
        while (i < m) {
            if (p.charAt(i) == p.charAt(len)) {
                len++;             // extend prefix-suffix length
                lps[i] = len;      // set lps[i]
                i++;
            } else {
                if (len != 0) {
                    len = lps[len - 1]; // fall back to previous lps
                } else {
                    lps[i] = 0;    // no prefix-suffix match
                    i++;
                }
            }
        }
        return lps;
    }

    // KMP search function
    static List<Integer> kmpSearch(String text, String pat) {
        int n = text.length();     // length of text
        int m = pat.length();      // length of pattern
        List<Integer> res = new ArrayList<>();
        if (m == 0) return res;    // empty pattern -> no matches

        int[] lps = buildLPS(pat); // precompute lps array
        int i = 0; // index for text
        int j = 0; // index for pattern

        while (i < n) {
            if (text.charAt(i) == pat.charAt(j)) {
                i++;
                j++;
                if (j == m) {      // full match found
                    res.add(i - j); // store start index
                    j = lps[j - 1]; // continue search using lps
                }
            } else {
                if (j != 0) {
                    j = lps[j - 1]; // use fallback in pattern
                } else {
                    i++;            // move in text
                }
            }
        }
        return res;
    }

    // Main function to demonstrate
    public static void main(String[] args) {
        String text = "ABABDABACDABABCABAB";
        String pattern = "ABABCABAB";

        System.out.println("Text:    " + text);
        System.out.println("Pattern: " + pattern);

        int[] lps = buildLPS(pattern);
        System.out.println("\nLPS Array: " + Arrays.toString(lps));

        List<Integer> matches = kmpSearch(text, pattern);
        System.out.println("\nMatch starting indices (0-based): " + matches);
        for (int idx : matches) {
            System.out.println("Match at index " + idx + ": '" 
                + text.substring(idx, idx + pattern.length()) + "'");
        }
    }
}

File: 09.KMPAlgo.py
# Knuth–Morris–Pratt (KMP) is a linear-time string matching algorithm that finds all occurrences of a pattern
# in a text by precomputing a longest-prefix-that-is-also-suffix table (LPS / failure function).
# It uses the LPS to skip unnecessary comparisons, achieving O(n + m) time
# where n = text length and m = pattern length.

# Explanation (intuition + how it works)

# LPS array (failure function) — for each pattern index i, lps[i] stores the
# length of the longest proper prefix of pattern[:i+1] that is also a suffix of pattern[:i+1].
# Proper = not whole string. This tells us where to resume matching in the pattern after a mismatch.

# Build LPS — iterate the pattern once using a two-pointer idea: i scans current position, length keeps track of current matched prefix length. On mismatch you fall back to lps[length-1] (like jumping to the previous possible prefix), never decreasing i unless necessary.

# Search loop — scan the text with i and the pattern with j. On match advance both. On mismatch:

# if j > 0, set j = lps[j-1] (resume from last safe position in pattern),

# else advance i.
# This avoids re-checking characters in text that we already matched.

# Complexity — building LPS is O(m), searching is O(n); total O(n + m). Extra space O(m) for LPS.

def build_lps(p):
    m = len(p)
    lps = [0] * m
    length = 0
    i = 1

    while i < m:
        if p[i] == p[length]:
            length += 1
            lps[i] = length
            i += 1
        else:
            if length != 0:
                length = lps[length - 1]
            else:
                lps[i] = 0
                i += 1
    return lps


def kmp_search(text, pattern):
    n = len(text)
    m = len(pattern)
    if m == 0:
        return []
    lps = build_lps(pattern)
    res = []
    i = 0
    j = 0
    while i < n:
        if text[i] == pattern[j]:
            i += 1
            j += 1
            if j == m:
                res.append(i - j)
                j = lps[j - 1]
        else:
            if j != 0:
                j = lps[j - 1]
            else:
                i += 1
    return res


text = "ABABDABACDABABCABAB"
pattern = "ABABCABAB"

print("Text: ", text)
print("Pattern: ", pattern)
print("\nBuilding LPS array for the pattern...")
lps = build_lps(pattern)
print("LPS array:", lps)
print("\nSearching for pattern occurrences...")
matches = kmp_search(text, pattern)
print("Match starting indices (0-based):", matches)
for idx in matches:
    print(f"Match at index {idx}: '{text[idx: idx + len(pattern)]}'")

